internal/transports/p2p/peer/checkpoint.go

```diff
@@ -0,0 +1,139 @@
+package peer
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/bitcoin-sv/block-headers-service/domains"
+	"github.com/bitcoin-sv/block-headers-service/internal/chaincfg"
+	"github.com/bitcoin-sv/block-headers-service/internal/chaincfg/chainhash"
+	"github.com/rs/zerolog"
+)
+
+type checkpoint struct {
+	checkpoint      *chaincfg.Checkpoint
+	currentIndex    int
+	finalCheckpoint *chaincfg.Checkpoint
+	checkpoints     []chaincfg.Checkpoint
+	log             *zerolog.Logger
+	lock            sync.RWMutex
+}
+
+func newCheckpoint(checkpoints []chaincfg.Checkpoint, tipHeight int32, log *zerolog.Logger) *checkpoint {
+	logger := log.With().Str("subservice", "checkpoint").Logger()
+	ch := &checkpoint{
+		checkpoints: checkpoints,
+		log:         &logger,
+	}
+
+	if len(checkpoints) != 0 {
+		ch.finalCheckpoint = &checkpoints[len(checkpoints)-1]
+		ch.next(tipHeight)
+	}
+
+	return ch
+}
+
+func (ch *checkpoint) Height() int32 {
```

Comment:
Since this is within `peer` package, we can either make these methods private, or move the checkpoint struct to another package within peer to have good encapsulation practice

Reply:
My intention was to clarify which methods you're supposed to use and which methods are private to that component. I agree that it can be made private; however, since the checkpoint struct is private and is a private field, it effectively remains private to other packages. In other words, I'm treating this as a way to create more access scopes. If you find this confusing, I can make it private. From my perspective, though, making it private could confuse devs (writing the peer code) about which methods are available for use.

Reply:
> From my perspective, though, making it private could confuse devs (writing the peer code) about which methods are available for use.
Yes, I agree with you @dorzepowski. Besides, IMHO, it's crucial to point out that using `ch.checkpoint.Height` is not "thread-safe" - instead of the public Height method. 

Maybe we could add a comment somewhere about it.